filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,272 @@
+	since time.Duration,
+	file string) (*Reader, error) {
+
+	args := []string{"--utc", "--output=json", "--follow"}
+	if file != "" && file != localSystemJournalID {
+		args = append(args, "--file", file)
+	}
+
+	args = handleSeekAndCursor(args, mode, since, cursor)
+
+	for _, u := range units {
+		args = append(args, "--unit", u)
+	}
+
+	for _, i := range syslogIdentifiers {
+		args = append(args, "--identifier", i)
+	}
+
+	for _, m := range matchers.Matches {
+		args = append(args, m.String())
+	}
+
+	for _, m := range transports {
+		args = append(args, fmt.Sprintf("_TRANSPORT=%s", m))
+	}
+
+	logger.Debugf("Journalctl command: journalctl %s", strings.Join(args, " "))
+	cmd := exec.Command("journalctl", args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	r := Reader{
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		errChan:  make(chan error),
+		logger:   logger,
+		stdout:   stdout,
+		stderr:   stderr,
+		canceler: canceler,
+	}
+
+	// Goroutine to read errors from stderr
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stderr goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stderr)
+		msgs := []string{}
+		for {
+			line, err := reader.ReadString('\n')
+			if errors.Is(err, io.EOF) {
+				if len(msgs) == 0 {
+					return
+				}
+				errMsg := fmt.Sprintf("Journalctl wrote errors: %s", strings.Join(msgs, "\n"))
+				logger.Errorf(errMsg)
+				r.errChan <- errors.New(errMsg)
+				return
+			}
+			msgs = append(msgs, line)
+		}
+	}()
+
+	// Goroutine to read events from stdout
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stdout goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stdout)
+		for {
+			data, err := reader.ReadBytes('\n')
+			if errors.Is(err, io.EOF) {
+				close(r.dataChan)
+				return
+			}
+
+			select {
+			case <-r.canceler.Done():
+				return
+			case r.dataChan <- data:
+			}
+		}
+	}()
+
+	if err := cmd.Start(); err != nil {
+		return &Reader{}, fmt.Errorf("cannot start journalctl: %w", err)
+	}
+
+	return &r, nil
+}
+
+// Close stops the `journalctl` process and waits for all
+// goroutines to return, the canceller passed to `New` should
+// be cancelled before `Close` is called
+func (r *Reader) Close() error {
+	if r.cmd == nil {
+		return nil
+	}
+
+	if err := r.cmd.Process.Kill(); err != nil {
+		return fmt.Errorf("cannot stop journalctl: %w", err)
+	}
+
+	r.logger.Debug("waiting for all goroutines to finish")
+	r.wg.Wait()
+	return nil
+}
+
+// Next returns the next journal entry. If there is no entry available
+// next will block until there is an entry or cancel is cancelled.
+//
+// If cancel is cancelled, Next returns a zero value JournalEntry
+// and ErrCancelled.
+func (r *Reader) Next(cancel input.Canceler) (JournalEntry, error) {
+	select {
```

Comment:
my filebeat knowledge is kind of limited, do we care if `Next()` blocks?

Reply:
Yes, we expect `Next()` to block until we have data to be converted into a Beat Event. Blocking on `Next()` is how the input waits for more data.